packages/stores/src/providers/ExchangeRatesProvider.tsx

```diff
@@ -1,52 +1,150 @@
-import React, { ReactNode } from 'react';
-import { useLocalStorage } from 'usehooks-ts';
-import { useSubscription } from '@deriv/api';
+import React from 'react';
+import { ExchangeRatesResponse } from '@deriv/api-types';
+import { useWS } from '@deriv/shared';
 import ExchangeRatesContext from '../stores/ExchangeRatesContext';
 
-type TExchangeRatesProvider = {
-    children: ReactNode;
+// Implement this type to prevent install @deriv/deriv-api to this package
+type DerivAPIBasicSubscribe = {
+    unsubscribe: () => void;
 };
-
+type TUnsubscribeFunction = (id: string) => void;
 type TRate = Record<string, Record<string, number>>;
+type TPayload = {
+    base_currency: string;
+    target_currency: string;
+};
+type TExchangeRatesProvider = {
+    children: React.ReactNode;
+};
 
 const ExchangeRatesProvider = ({ children }: TExchangeRatesProvider) => {
-    const [exchange_rates, setExchangeRates] = useLocalStorage<TRate>('exchange_rates', {});
+    const WS = useWS();
+    const [exchangeRates, setExchangeRates] = React.useState<TRate>({});
+    const isMounted = React.useRef(false);
+    const subscriptions = React.useRef<Record<string, DerivAPIBasicSubscribe>>();
+    const subscriptionsId = React.useRef<Record<string, string>>();
+    const exchangeRatesSubscriptions = React.useRef<string[]>([]);
 
-    const { subscribe, data, ...rest } = useSubscription('exchange_rates');
+    const subscribeToCurrencyRate = React.useCallback(
+        async (payload: TPayload) => {
+            const id = JSON.stringify(payload);
```

Comment:
future PRs, for building keys, better to use actual object parameters, e.g. source and target currency. As right now, to unsubscribe, one actually need the original payload of request (which typically is not practical, and prone to errors e.g. even due to different object keys order)
![Uploading Screenshot 2024-03-11 at 10.50.15.png…]()

Reply:
@wojciech-deriv Honestly, I don't understand what do you mean about `actual object parameters`? Right now the key is serialized object with `base_currency` and `target_currency`. Do you suggest to use just `target_currency` as a key?